### PR TITLE
SALTO-2156: Fix unneeded cache invalidation in multi-env source

### PR DESCRIPTION
**Pull request by ori-moisis**
Before this change, if there was no semantic change in a nacl file source, the other caches
(multi env and workspace) would not be updated, which meant the hash would become out of sync
leading to incorrect cache invalidation the next time a semantic change is detected.

The fix here keeps the hash as a non-semantic meaning (unfortunately) which means there are still
many different hashes that can all represent to same elements.
this is not ideal, but hopefully now the nacl source and multi env source would both be updated
so they do not lose sync of which hash is the current hash.

---

Note that this re-calculation of the cache, when it happened in "updateNaclFiles", would cause us to report
an incorrect number of changes. so, if this cache invalidation would happen on fetch for example, we would
report a number of changes as if the entire workspace was changed.

---
_Release Notes_: 
- Fixed issue where sometimes fetch would take longer than needed and report the wrong number of changes

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 1]
```diff
@@ -216,18 +216,14 @@ export const createMergeManager = async (flushables: Flushable[],
...
-    await hashes.set(getMergedHashKey(cacheUpdate.src1Prefix, cacheUpdate.src2Prefix),
-      postChangeHash)
-    const cacheValid = ((!preChangeHash && !cachePreChangeHash)
-      || preChangeHash === cachePreChangeHash)
-      && src1Changes.cacheValid && src2Changes.cacheValid
+    const preChangeHash = (src1Changes.preChangeHash || '') + (src2Changes.preChangeHash || '')
+    const mergedHashKey = getMergedHashKey(cacheUpdate.src1Prefix, cacheUpdate.src2Prefix)
+    const cachePreChangeHash = await hashes.get(mergedHashKey) ?? ''
+    const cacheValid = (
+      preChangeHash === cachePreChangeHash
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:223]: this condition changed
it used to be
```
(!preChangeHash && !cachePreChangeHash) || preChangeHash === cachePreChangeHash
```

it took me a while to read and understand it, and I really could not figure out why the first part was needed so I removed it

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 2]
```diff
@@ -215,12 +215,9 @@ const buildMultiEnvSource = (
...
   Promise<{ state: MultiEnvState; changes: EnvsChanges }> => {
-    const states: Record<string, SingleState> = Object.fromEntries(
-      (await awu(Object.keys(sources))
-        .filter(name => name !== commonSourceName)
-        .map(async name => [name, state?.states[name]
-          ?? (await buildStateForSingleEnv(name))])
-        .toArray())
+    const states = await mapValuesAsync(
+      envSources(),
+      async (_source, envName) => state?.states[envName] ?? buildStateForSingleEnv(envName)
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:220]: there should be no logic change here, only style change

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 3]
```diff
@@ -237,15 +234,6 @@ const buildMultiEnvSource = (
       states,
       mergeManager,
     }
-    if (Object.values(envChanges).every(changeSet => changeSet.changes.length === 0)) {
-      return { state: current, changes: {} }
-    }
-    const changesInCommon = (envChanges[commonSourceName]
-      ?.changes ?? []).length > 0
-    const relevantEnvs = Object.keys(sources)
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:245]: this is the main bug fix - removing `relevantEnvs` to ensure we always call `mergeComponents` for all environments (`mergeComponents` should ensure the "caches" remain in-sync)

File: packages/workspace/test/workspace/multi_env/multi_env_source.test.ts [hunk 4]
```diff
@@ -197,13 +199,165 @@ const source = multiEnvSource(
...
+  maps: Record<string, InMemoryRemoteMap<unknown>>
+  creator: RemoteMapCreator
+}
+const mockRemoteMaps = (): MockRemoteMapCreator => {
+  const maps: Record<string, InMemoryRemoteMap<unknown>> = {}
+  return {
+    maps,
+    creator: async <T, K extends string>({ namespace }: CreateRemoteMapParams<T>) => {
+      maps[namespace] = maps[namespace] ?? new InMemoryRemoteMap()
+      return maps[namespace] as unknown as RemoteMap<T, K>
```
Comment (ori-moisis) [inline packages/workspace/test/workspace/multi_env/multi_env_source.test.ts:212]: could not figure out a way to avoid this cast :(

File: packages/workspace/test/workspace/multi_env/multi_env_source.test.ts [hunk 5]
```diff
@@ -197,13 +199,165 @@ const source = multiEnvSource(
...
+  }
+}
 
 describe('multi env source', () => {
   let referencedByIndex: RemoteMap<ElemID[]>
   beforeAll(async () => {
     await source.load({})
     referencedByIndex = createMockRemoteMap<ElemID[]>()
   })
+  describe('load', () => {
```
Comment (ori-moisis) [inline packages/workspace/test/workspace/multi_env/multi_env_source.test.ts:223]: this does not cover the interface of `load` by any means.
I added a few basic scenarios in addition to the one I fixed because it was completely untested, but adding real test coverage for this I think is out of scope for this bug fix

File: packages/workspace/src/workspace/nacl_files/parsed_nacl_file.ts [hunk 6]
```diff
@@ -18,7 +18,7 @@ import { SourceMap, ParseError } from '../../parser'
 
 export type ParsedNaclFileData = {
   errors: () => Promise<ParseError[] | undefined>
-  referenced: () => Promise<string[] | undefined>
+  referenced: () => Promise<string[]>
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/parsed_nacl_file.ts:21]: Can't we change this for the errors as well?

**Review by roironn — APPROVED** [review]

File: packages/workspace/src/workspace/nacl_files/parsed_nacl_file.ts (hunk 6 above)
Comment (ori-moisis) [reply packages/workspace/src/workspace/nacl_files/parsed_nacl_file.ts:21]: looks like there is some intentional behavior that I don't understand around errors
https://github.com/salto-io/salto/pull/2456
I could not find sufficient context as to why the above change was needed, so I am not going to touch the errors logic.
